proxy/round_tripper/proxy_round_tripper.go

```diff
@@ -233,13 +236,39 @@ func (rt *roundTripper) backendRoundTrip(
 
 	rt.combinedReporter.CaptureRoutingRequest(endpoint)
 	tr := GetRoundTripper(endpoint, rt.roundTripperFactory)
-	res, err := tr.RoundTrip(request)
+	res, err := rt.timedRoundTrip(tr, request)
 
 	// decrement connection stats
 	iter.PostRequest(endpoint)
 	return res, err
 }
 
+func (rt *roundTripper) timedRoundTrip(tr http.RoundTripper, request *http.Request) (*http.Response, error) {
+
+	if rt.endpointTimeout <= 0 {
+		return tr.RoundTrip(request)
+	}
+	reqCtx, cancel := context.WithTimeout(request.Context(), rt.endpointTimeout)
+	request = request.WithContext(reqCtx)
+	errored := make(chan struct{}, 1)
+
+	go func() {
```

Comment:
@swetharepakula could you explain the purpose of this `goroutine` and the `select`? It currently appears unnecessary.

Reply:
@zachgersh, the need for the goroutine is for the `defer cancel()`. We need to make sure that cancel gets called in all situations.

Reply:
@zachgersh @swetharepakula I think this code should look something like this:

```go
reqCtx, cancel := context.WithTimeout(request.Context(), rt.endpointTimeout)
defer cancel()
```

There doesn't appear to be a need for the goroutine or `errored` channel, and the `if err != nil` block can be simplified.

`RoundTrip` already handles `ctx.Err()` and returns it, so I don't think there's a need to check this.

Reply:
The need for the goroutine is because the function will exit before the complete response is returned. `tr.RoundTrip` returns as soon as the headers start coming in for the response, but it does not mean the response is completed. So if `defer cancel()` is added then the request is canceled as soon as the headers start coming in.
Here is the discussion about this particular PR: https://cloudfoundry.slack.com/archives/C05586PBX/p1520911938000018?thread_ts=1520893756.000433&cid=C05586PBX

Reply:
@swetharepakula @zachgersh Looking at this some more, I don't think `cancel()` ever gets called until the timeout elapses unless there's an error from `tr.RoundTrip`. While you're not leaking contexts they certainly hang around a while.

It looks like the context is created too late in the request pipeline, that's why you had to jump through hoops to get this to work. You should be able to call `cancel()` as soon as you're done with a context, not after the timeout fires.

Reply:
@judwhite I agree that this *looks* overly complicated.  We just haven't figured out how to do it better.  Pull requests are welcome though....